### PR TITLE
added docker-compose support

### PR DESCRIPTION
**Pull request by smoebody**
with this one can use docker and docker-compose to set up a simple dev-environment, means not having nodejs on his computer.

    docker-compose run --rm node npm install

calls `npm install` inside the docker container

    docker-compose run --rm node

calls `npm test` - the default command - and runs all gulp-tasks. 

    docker-compose run --rm --service-ports node node --debug-brk node_modules/.bin/gulp

calls the gulp-script with node's debug-option. connect your debug-client to localhost:5858.

The option `--rm` ensures that the container is thrown away after command exits.

### CODE REVIEW

**Comment by whitlockjc** [comment]
Thanks for your help.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+version: '2'
+services:
+  node:
+    image: docker.io/smoebody/dev-nodejs:6
```
Comment (imZack) [inline docker-compose.yml]: I'm curious about why not just use official node image?

**Comment by whitlockjc** [comment]
Good point, I didn't pay attention.  We should use the alpine-node or official node image.  If @smoebody doesnt submit a follow-up PR, I'll fix it.

**Comment by smoebody** [comment]
im not using the official node image because all commands are run as root. my image creates a user `dev` which inherits the uid and gid of the user the owner of the project folder. by running all scripts as this user ensures that the project owner is able to delete files created by the docker-process, e.g. node_modules*

my image uses the official node-images though. right now im on it to publish the sources and set up automated docker-builds as well as a proper documentation. if thats the problem of using a non-official image just give me a minute.

**Comment by imZack** [comment]
How about using `--user` like and you will be able to delete files created in container?
```
docker run --user `id -u`
```

docs: https://docs.docker.com/engine/reference/run/#/user

**Comment by smoebody** [comment]
ok, i wasn't aware of that. i will provide a follow-up PR to fix that

**Comment by smoebody** [comment]
@imZack do you know how to use that within the `docker-compose.yml`? i cannot get it to work.
if not, i suggest to remove the docker-compse.yml

**Comment by imZack** [comment]
Updated docker-compose.yml
```
version: '2'
services:
  node:
    image: node:alpine
    user: ${UID}
    working_dir: /usr/src/app
    volumes:
      - ./:/usr/src/app:z
      - ./npm-cache:/.npm:z
    ports:
      - 127.0.0.1:5858:5858
    environment:
      - NODE_ENV=development
    command:
      - npm test
```

And commands like below:
```
UID=`id -u` docker-compose run --rm node npm install
UID=`id -u` docker-compose run --rm node npm test
```

I think this should work.

**Comment by whitlockjc** [comment]
Do you mind creating a PR for this?  Also, why Docker Compose and not just a `Dockerfile`?

**Comment by smoebody** [comment]
i can't get it to run with the docker-compose.yml above.
```
$ UID=`id -u` docker-compose run --rm node npm install                                                                                                                         
-bash: UID: readonly variable
WARNING: The UID variable is not set. Defaulting to a blank string.
```
which leads to the script run as *root*

when i change the variable name to something like `${USER_ID}` i can start the script but `npm install` tries to create the cache-folder in the containers root-folder because of the absence of an user-folder. he obviously has no rights to write, only way is to create the `npm-cache`-folder in advance as project owner, which leads to having this folder checked in to the repository which i think is not a good idea. 

if we cannot use the practices docker provides we should consider not using it.

so there are three options right now:

1. use my image
1. use the official image and run the scripts as root, which leads to files in project-folder you need `sudo` for
1. don't use docker at all

@whitlockjc: what do you mean by *using a Dockerfile*? in my understanding you only need a Dockerfile if you want to create your own image, which you should only do if no other images satisfy your needs or if you want to provide something other images don't.

**Comment by whitlockjc** [comment]
So you're basically just using Docker Compose to run `swagger-tools` but not for building an image to distribute?  I'm beginning to think that the Docker Compose support in `swagger-tools` might not make sense.

**Comment by smoebody** [comment]
I use docker to set up a development testing environment. It is meant to provide nodejs and npm to run the tests and build script in a constant environment.
I do not have all the nodejs versions on my machine. In fact i only have the version distibuted by my linux distro. Docker enables me to instantly switch environments. 

Thats my purpose to provide a docker-compose.yml. Yes, Im a docker fanboy :-) 

Am 12. Dezember 2016 18:44:27 MEZ, schrieb Jeremy Whitlock <notifications@github.com>:
>So you're basically just using Docker Compose to run `swagger-tools`
>but not for building an image to distribute?  I'm beginning to think
>that the Docker Compose support in `swagger-tools` might not make
>sense.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/apigee-127/swagger-tools/pull/458#issuecomment-266499057

free art - free mind

**Comment by whitlockjc** [comment]
I use [nvm](https://github.com/creationix/nvm) for that myself but Docker could just as easily be used to do the same.  My concern is that if our Docker support is not generic, we probably shouldn't have it.  If the plan is just to be able to run `gulp test` or what not within Docker, you could just use the mainstream node images and mount your `swagger-tools` source directory in the container and run whatever `node`/`npm` command you need in it.

I think I'm going to delete the Docker Compose support and instead document how to run tests within Docker.  Would that work?

**Comment by smoebody** [comment]
Im fine with that.

Am 12. Dezember 2016 19:15:01 MEZ, schrieb Jeremy Whitlock <notifications@github.com>:
>I use [nvm](https://github.com/creationix/nvm) for that myself but
>Docker could just as easily be used to do the same.  My concern is that
>if our Docker support is not generic, we probably shouldn't have it. 
>If the plan is just to be able to run `gulp test` or what not within
>Docker, you could just use the mainstream node images and mount your
>`swagger-tools` source directory in the container and run whatever
>`node`/`npm` command you need in it.
>
>I think I'm going to delete the Docker Compose support and instead
>document how to run tests within Docker.  Would that work?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/apigee-127/swagger-tools/pull/458#issuecomment-266507085

free art - free mind

**Pull request merged**